src/coreclr/vm/object.cpp

```diff
@@ -1494,7 +1494,7 @@ void StackTraceArray::EnsureThreadAffinity()
         StackTraceArray copy;
         GCPROTECT_BEGIN(copy);
             copy.CopyFrom(*this);
-            this->Swap(copy);
+            this->Set(copy.Get());
         GCPROTECT_END();
     }
 }
```

Comment:
The `StackTraceArray` has a header (`StackTraceArray::ArrayHeader`) containing a m_size member that's updated as the last item when the array is modified (see `StackTraceArray::Append`). Since readers only read upto "size" number of elements, there should be no torn reads possible.

Reply:
This works as long as the array is not reused and overwritten.

I see comments like https://github.com/dotnet/runtime/pull/100427/files#diff-47cbfa47fb88845e5e76fd4df0a84e9da494508ffe80da116511fc2d0becd9baR2248 that make it sound like that there are situations where the array is reused, but they just may be stale or misleading.

Reply:
It is never reused or overwritten. It is either appended to or replaced by a brand new one - either by an empty one or by a new one with a copy of the contents of the old one, but only the `m_size` elements are copied. 
I wonder which of the comments you are referring to, the link has shown me just the whole diff without anything specific marked.

Reply:
> I wonder which of the comments you are referring to

```
// Cleanup previous info
gc.stackTrace.Append(m_pStackTrace, m_pStackTrace + m_dFrameCount);
```

Reply:
Ah, I can see how this is misleading. The gc.StackTrace is empty at this point (the m_array is NULL), the Append results in allocating a new managed array and copying the frames into it. And in the end, the gc.stackTrace is written to the Exception, essentially clearing what was there before and creating a new stack trace from scratch.